### PR TITLE
Upgrade to yaml-cpp 0.8.0

### PR DESCRIPTION
**Pull request by marcoag**
Upgrade to 0.8.0 and some adjustments to the patch.

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]
According with the [noble yaml-cpp package documentation](https://packages.ubuntu.com/source/noble/yaml-cpp) this is the right version

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20714/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15024/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21399/)

**Comment by marcoag** [comment]
Another round of CI, without the patch:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20788/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15088/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21466/)

**Review by ahcorde — CHANGES_REQUESTED** [review]
https://ci.ros2.org/job/ci_windows/21466/console

```
19:03:28 CMake Error at C:/ci/ws/install/share/ament_cmake_vendor_package/cmake/ament_vendor.cmake:157 (message):
19:03:28   ament_vendor() could not find patch file:
19:03:28   C:/ci/ws/src/ros2/yaml_cpp_vendor/patches
19:03:28 Call Stack (most recent call first):
19:03:28   C:/ci/ws/install/share/ament_cmake_vendor_package/cmake/ament_vendor.cmake:107 (_ament_vendor)
19:03:28   CMakeLists.txt:19 (ament_vendor)
```

**Comment by clalancette** [comment]
Here's CI again, with the patches argument removed:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20790/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15090/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21468/)

**Review by ahcorde — CHANGES_REQUESTED** [review]
```
21:37:16 C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\include\rosbag2_storage/yaml.hpp(28,10): fatal error C1083: Cannot open include file: 'yaml-cpp/yaml.h': No such file or directory (compiling source file C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\src\rosbag2_storage\storage_options.cpp) [C:\ci\ws\build\rosbag2_storage\rosbag2_storage.vcxproj]

21:37:16 C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\include\rosbag2_storage/yaml.hpp(28,10): fatal error C1083: Cannot open include file: 'yaml-cpp/yaml.h': No such file or directory (compiling source file C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\src\rosbag2_storage\storage_factory.cpp) [C:\ci\ws\build\rosbag2_storage\rosbag2_storage.vcxproj]

21:37:16 C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\include\rosbag2_storage/yaml.hpp(28,10): fatal error C1083: Cannot open include file: 'yaml-cpp/yaml.h': No such file or directory (compiling source file C:\ci\ws\src\ros2\rosbag2\rosbag2_storage\src\rosbag2_storage\base_io_interface.cpp) [C:\ci\ws\build\rosbag2_storage\rosbag2_storage.vcxproj]
```

**Comment by clalancette** [comment]
All right, so I know what the problem is.  In `yaml-cpp` version 0.8.0, they actually fixed it so that it exposes a proper CMake-style target, i.e. `yaml-cpp::yaml-cpp`.  Prior to version 0.8.0, they were using just `yaml-cpp`.  That's why Windows can't find it in this CI.

To fix this, we are going to have to have additional PRs targeting the rosbag2 repository, the rviz repository, and the image_common repositories changing things to use `yaml-cpp::yaml-cpp`.

**Comment by clalancette** [comment]
Here is CI including the new PRs:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20811/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15115/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21494/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/21495/)
* RHEL-9 [Build Status](http://ci.ros2.org/job/ci_linux-rhel/632/)

File: yaml_cpp_vendor-extras.cmake.in [hunk 1]
```diff
@@ -5,21 +5,3 @@ else()
...
 message(STATUS "Setting yaml-cpp_DIR to: '${yaml-cpp_DIR}'")
-
-find_package(yaml-cpp REQUIRED)
-
-set(yaml_cpp_vendor_LIBRARIES ${YAML_CPP_LIBRARIES})
-set(yaml_cpp_vendor_INCLUDE_DIRS ${YAML_CPP_INCLUDE_DIR})
-
-# since yaml-cpp 0.8.0, yaml-cpp supports the yaml-cpp::yaml-cpp target
-if(TARGET yaml-cpp::yaml-cpp)
-  list(APPEND yaml_cpp_vendor_TARGETS yaml-cpp::yaml-cpp)
```
Comment (981213) [inline yaml_cpp_vendor-extras.cmake.in:16]: Hello @clalancette !
@SteveMacenski pointed out in https://github.com/ros-planning/navigation2/pull/4267 that the yaml_cpp_vendor package should have pulled in the dependencies itself.
Why is this dropped in favor of changing CMakeLists.txt in every downstream packages that need yaml-cpp?

Comment (clalancette) [reply yaml_cpp_vendor-extras.cmake.in:16]: > @SteveMacenski pointed out in [ros-planning/navigation2#4267](https://github.com/ros-planning/navigation2/pull/4267) that the yaml_cpp_vendor package should have pulled in the dependencies itself.
> Why is this dropped in favor of changing CMakeLists.txt in every downstream packages that need yaml-cpp?

So there are 2 reasons:

1.  When I originally did this, I couldn't actually get this snippet of code to work.  However, when I started debugging it I realized that...
2. Downstream packages should actually do `find_package(yaml-cpp REQUIRED)`.  That matches what we do in other vendor packages, like [spdlog_vendor](https://github.com/ros2/spdlog_vendor), and things that depend on it (https://github.com/ros2/rcl_logging/blob/128d88e6e4ad6df97405564afbdffa2ebc167814/rcl_logging_spdlog/CMakeLists.txt#L19-L20).  The thinking here is that having a vendor package is a "crutch", and someday in the future we may remove it.  Doing `find_package(spdlog_vendor REQUIRED)` thus should only make the paths available, and the downstream packages should indeed always have to do `find_package(spdlog REQUIRED)` to guard against the day where we no longer provide the vendor package.

That said, I understand that this is a change, so if you need me to open a PR to downstream somewhere, let me know and I can do that.

Comment (SteveMacenski) [reply yaml_cpp_vendor-extras.cmake.in:16]: Wait - but why wouldn't the vendor include all the things of the software its vendorizing itself? That seems to break common sense encapsulation, no? If we were to require both, wouldn't it be sensible then that the vendorized packages are only in the package.xml to fetch the compiled resources & then in the CMakeLists we're only finding the main package (ie don't `find_package(*_vendor)`? What's the value that the vendorized package provides from the cmake side of things if we have to include the main package separately?

Comment (clalancette) [reply yaml_cpp_vendor-extras.cmake.in:16]: > What's the value that the vendorized package provides from the cmake side of things if we have to include the main package separately?

The value is mostly in the "build-if-needed" part, which we utilize on Windows.  The honest truth is that if we didn't have Windows, most of these vendor packages wouldn't exist and we'd just use them from the system.  And someday in the future when we use conda (or similar) to install our dependencies on Windows, we will remove these vendor packages.

> If we were to require both, wouldn't it be sensible then that the vendorized packages are only in the package.xml to fetch the compiled resources & then in the CMakeLists we're only finding the main package (ie don't `find_package(*_vendor)`?

That would be nice.  I remember we've talked about this before, and there was some reason we couldn't or didn't want to just `find_package(yaml-cpp)` and not the vendor package.  Maybe @cottsay can comment more here.

**Pull request merged**